language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Ck.hs

```diff
@@ -19,69 +22,178 @@ import           PlutusPrelude
 
 import           Language.PlutusCore.Constant
 import           Language.PlutusCore.Core
+import           Language.PlutusCore.Error
 import           Language.PlutusCore.Evaluation.Machine.ExBudgeting
-import           Language.PlutusCore.Evaluation.Machine.ExBudgetingDefaults
+import           Language.PlutusCore.Evaluation.Machine.ExBudgetingDefaults (defaultCostModel)
 import           Language.PlutusCore.Evaluation.Machine.Exception
-import           Language.PlutusCore.Evaluation.Machine.ExMemory            (Plain)
 import           Language.PlutusCore.Evaluation.Result
 import           Language.PlutusCore.Name
-import           Language.PlutusCore.Pretty                                 (PrettyConst)
+import           Language.PlutusCore.Pretty                                 (PrettyConfigPlc, PrettyConst)
 import           Language.PlutusCore.Universe
-import           Language.PlutusCore.View
 
+import           Control.Monad.Reader
 import           Data.Array
+import qualified Data.Map                                                   as Map
 
 infix 4 |>, <|
 
--- | The CK machine throws this error when it encounters a 'DynBuiltinName'.
-data NoDynamicBuiltinNamesMachineError
-    = NoDynamicBuiltinNamesMachineError
+data CkValue uni =
+    VCon (Term TyName Name uni ())  -- TODO: Really want a constant here.
+  | VTyAbs TyName (Kind ()) (Term TyName Name uni ())
+  | VLamAbs Name (Type TyName uni ()) (Term TyName Name uni ())
+  | VIWrap (Type TyName uni ()) (Type TyName uni ()) (CkValue uni)
+  | VBuiltin
+      BuiltinName
+      Arity                -- Sorts of arguments to be provided (both types and terms): *don't change this*.
+      Arity                -- A copy of the arity used for checking applications/instantiatons: see Note [Arities in VBuiltin]
+      [Type TyName uni ()] -- The types the builtin is to be instantiated at.
+                           -- We need these to construct a term if the machine is returning a stuck partial application.
+      [CkValue uni]        -- Arguments we've computed so far.
+    deriving (Show, Eq)   -- Eq is just for tests.
+
+mkBuiltinApplication :: () -> BuiltinName -> Arity -> [Type TyName uni ()] -> [Term TyName Name uni ()] -> Term TyName Name uni ()
```

Comment:
This *is* the same as the CEK one if you abstract over `ann`, right?

Reply:
Ah, you're right!  I wasn't paying attention to what I was writing.  That does give us more opportunities for sharing code...

Reply:
Wait, they're _not_ the same after all.  Some of the Values contain terms, and in because of this `CekValue`s have to contain environments binding the free variables.  In the CK machine all of the terms are supposedly closed, so we just have terms without environments.  However, I now realise that we have to be sure that all of the terms in a `CkValue` are closed.  I'll have to think about that for a bit to convince myself that we're OK.